File: 2025/day/7/Main.hs
{-# LANGUAGE MultilineStrings #-}

import Data.Array
import Data.List

part1 :: Int -> [[Int]] -> Int
part1 s = snd . foldl f ([s], 0)
  where
    f (beams, numSplits) splitters =
      let (splittingBeams, passthroughBeams) = partition (`elem` splitters) beams
          splitBeams = [x + dx | x <- splittingBeams, dx <- [-1, 1]]
          beams' = nub $ sort (passthroughBeams ++ splitBeams)
       in (beams', numSplits + length splittingBeams)

part2 :: Int -> Array Int [Int] -> Integer
part2 x splitterIndices = memo ! (x, 0)
  where
    (_, ymax) = bounds splitterIndices
    xmax = foldr (max . maximum) 0 (elems splitterIndices)
    bnds = ((0, 0), (xmax + 1, ymax + 1))
    memo = array bnds [(xy, subproblem xy) | xy <- range bnds]
    subproblem (x, y)
      | y >= ymax + 1 = 1
      | x `elem` (splitterIndices ! y) = memo ! (x - 1, y + 1) + memo ! (x + 1, y + 1)
      | otherwise = memo ! (x, y + 1)

main = do
  input <- getContents
  let startLine : splitterLines = lines input
      Just s = elemIndex 'S' startLine
      splitterIndices =
        [ elemIndices '^' line
        | line <- splitterLines,
          not (all (== '.') line) -- Remove void lines
        ]

  -- Part 1
  print $ part1 s splitterIndices

  -- Part 2
  let splitterIndicesArr = listArray (0, length splitterIndices - 1) splitterIndices
  print $ part2 s splitterIndicesArr

input =
  """
  .......S.......
  ...............
  .......^.......
  ...............
  ......^.^......
  ...............
  .....^.^.^.....
  ...............
  ....^.^...^....
  ...............
  ...^.^...^.^...
  ...............
  ..^...^.....^..
  ...............
  .^.^.^.^.^...^.
  ...............
  """

File: 2025/day/7/main.py
from functools import cache


input = """.......S.......
...............
.......^.......
...............
......^.^......
...............
.....^.^.^.....
...............
....^.^...^....
...............
...^.^...^.^...
...............
..^...^.....^..
...............
.^.^.^.^.^...^.
..............."""

# Read lines, strip "void" lines
def void(line):
    return all(c == "." for c in line)

lines = [line for line in input.splitlines() if not void(line)]

with open(0) as f:
    lines = [line.rstrip("\n") for line in f if not void(line.rstrip("\n"))]

S = next(i for i, c in enumerate(lines[0]) if c == "S")

# Parse each line to find splitters
splitters_per_line = [{i for i, c in enumerate(line) if c == "^"} for line in lines[1:]]

# Part 1
beams = {S}
num_splits = 0

for splitters in splitters_per_line:
    splitting_beams = beams & splitters
    passthrough_beams = beams - splitters
    split_beams = {x + dx for x in splitting_beams for dx in [-1, 1]}
    beams = passthrough_beams | split_beams
    num_splits += len(splitting_beams)

print(num_splits)

# Part 2
@cache
def dfs(line_no, beam_x):
    if line_no >= len(lines):
        return 0
    
    if line_no == len(lines) - 1:
        return 1

    if beam_x in splitters_per_line[line_no]:
        return dfs(line_no + 1, beam_x - 1) + dfs(line_no + 1, beam_x + 1)
    else:
        return dfs(line_no + 1, beam_x)

print(dfs(0, S))